network/connecteventmanager.go

```diff
@@ -11,96 +11,198 @@ type ConnectionListener interface {
+	done        chan struct{}
 }
 
-type connState struct {
-	refs       int
-	responsive bool
+type peerState struct {
+	newState, curState state
+	pending            bool
 }
 
 func newConnectEventManager(connListener ConnectionListener) *connectEventManager {
-	return &connectEventManager{
+	evtManager := &connectEventManager{
 		connListener: connListener,
-		conns:        make(map[peer.ID]*connState),
+		peers:        make(map[peer.ID]*peerState),
+		done:         make(chan struct{}),
 	}
+	evtManager.cond = sync.Cond{L: &evtManager.lk}
+	return evtManager
 }
 
-func (c *connectEventManager) Connected(p peer.ID) {
+func (c *connectEventManager) Start() {
+	go c.worker()
+}
+
+func (c *connectEventManager) Stop() {
 	c.lk.Lock()
-	defer c.lk.Unlock()
+	c.stop = true
+	c.lk.Unlock()
+	c.cond.Broadcast()
+
+	<-c.done
+}
+
+func (c *connectEventManager) getState(p peer.ID) state {
+	if state, ok := c.peers[p]; ok {
+		return state.newState
+	} else {
+		return stateDisconnected
+	}
+}
 
-	state, ok := c.conns[p]
+func (c *connectEventManager) setState(p peer.ID, newState state) {
+	state, ok := c.peers[p]
 	if !ok {
-		state = &connState{responsive: true}
-		c.conns[p] = state
+		state = new(peerState)
+		c.peers[p] = state
+	}
+	state.newState = newState
+	if !state.pending && state.newState != state.curState {
+		state.pending = true
+		c.changeQueue = append(c.changeQueue, p)
+		c.cond.Broadcast()
+	}
+}
+
+func (c *connectEventManager) worker() {
+	c.lk.Lock()
+	defer c.lk.Unlock()
+	defer close(c.done)
+
+	for {
+		for !c.stop && len(c.changeQueue) == 0 {
+			c.cond.Wait()
+		}
+
+		if c.stop {
+			return
+		}
+
+		pid := c.changeQueue[0]
+		c.changeQueue[0] = peer.ID("")
+		c.changeQueue = c.changeQueue[1:]
+
+		state, ok := c.peers[pid]
+		// If we've disconnected and forgotten, continue. We shouldn't reach this?
+		if !ok {
+			continue
+		}
+
+		// Record the fact that this "state" is no longer in the queue.
+		state.pending = false
+
+		// Then, if there's nothing to do, continue.
+		if state.curState == state.newState {
+			continue
+		}
+
+		// Or record the state update, then apply it.
+		oldState := state.curState
+		state.curState = state.newState
+
+		switch state.newState {
+		case stateDisconnected:
+			delete(c.peers, pid)
+			fallthrough
+		case stateUnresponsive:
+			// Only trigger a disconnect event if the peer was responsive.
+			// We could be transitioning from unresponsive to disconnected.
+			if oldState == stateResponsive {
+				c.lk.Unlock()
+				c.connListener.PeerDisconnected(pid)
+				c.lk.Lock()
+			}
+		case stateResponsive:
+			c.lk.Unlock()
+			c.connListener.PeerConnected(pid)
+			c.lk.Lock()
+		}
 	}
-	state.refs++
+}
+
+// Called whenever we receive a new connection. May be called many times.
+func (c *connectEventManager) Connected(p peer.ID) {
+	c.lk.Lock()
+	defer c.lk.Unlock()
 
-	if state.refs == 1 && state.responsive {
-		c.connListener.PeerConnected(p)
+	// disconnected -> responsive
```

Comment:
I'm mostly just replicating the current logic. But yes, I believe that's it.

But I'm also happy to change it to mark the peer responsive on any new connection. Honestly, that might be the better option because the peer might have been unresponsive because they were actually disconnected (and we just didn't know).

Thoughts?

Reply:
hard to say, it might be better to mark responsive on new connections. If the peer is sending out requests but not responding to them IIUC if we mark responsive whenever we receive a message from a peer then that message could also just be a request.

This seems to indicate we're not really doing ourselves much service here by not marking them as responsive, but I don't feel too strongly either way as long as we document what we're doing/why so people can figure out what we were up to when they figure out we're wrong and want to change it 🙃

Reply:
Eh, let's mark them as responsive. That seems reasonable.

Reply:
Done.